test/e2e/upgrade/upgrade_test.go

```diff
@@ -0,0 +1,87 @@
+package upgrade
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+	"testing"
+
+	"github.com/iancoleman/strcase"
+	v1 "github.com/infinispan/infinispan-operator/pkg/apis/infinispan/v1"
+	ispnctrl "github.com/infinispan/infinispan-operator/pkg/controller/infinispan"
+	tutils "github.com/infinispan/infinispan-operator/test/e2e/utils"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	CrdPath = "../../../deploy/crds/"
+)
+
+var (
+	testKube         = tutils.NewTestKubernetes(os.Getenv("TESTING_CONTEXT"))
+	upgradeStateFlow = []v1.ConditionType{v1.ConditionUpgrade, v1.ConditionStopping, v1.ConditionWellFormed}
+)
+
+// TODO remove and replace with DefaultSpec once Batch PR has been merged as it removes this logic from main_test
+var MinimalSpec = v1.Infinispan{
+	TypeMeta: tutils.InfinispanTypeMeta,
+	ObjectMeta: metav1.ObjectMeta{
+		Name: tutils.DefaultClusterName,
+	},
+	Spec: v1.InfinispanSpec{
+		Replicas: 1,
+	},
+}
+
+func TestGracefulShutdown(t *testing.T) {
+	namespace := tutils.Namespace
+	testKube.NewNamespace(namespace)
+	// Utilise the sha256 of the current Infinispan image to trick the operator into thinking a upgrade is required
+	oldImage := getDockerImageSha()
+	tutils.ExpectNoError(os.Setenv("DEFAULT_IMAGE", oldImage))
+	stopCh := testKube.InstallAndRunOperator(namespace, CrdPath, false)
+	ispn := MinimalSpec.DeepCopy()
+	ispn.Name = strcase.ToKebab(t.Name())
+	testKube.CreateInfinispan(ispn, namespace)
+	testKube.WaitForInfinispanPods(1, tutils.SinglePodTimeout, ispn.Name, namespace)
+	assertPodImage(oldImage, ispn)
+	close(stopCh)
+
+	// Unset DEFAULT_IMAGE and install the current operator version
+	os.Unsetenv("DEFAULT_IMAGE")
+	stopCh = testKube.InstallAndRunOperator(namespace, CrdPath, false)
+	defer close(stopCh)
+	for _, state := range upgradeStateFlow {
+		testKube.WaitForInfinispanCondition(ispn.Name, namespace, state)
+	}
+	assertPodImage(tutils.ExpectedImage, ispn)
+}
+
+func assertPodImage(image string, ispn *v1.Infinispan) {
+	pods := &corev1.PodList{}
+	err := testKube.Kubernetes.ResourcesList(ispn.Namespace, ispnctrl.PodLabels(ispn.Name), pods)
+	tutils.ExpectNoError(err)
+	for _, pod := range pods.Items {
+		if pod.Spec.Containers[0].Image != image {
+			tutils.ExpectNoError(fmt.Errorf("upgraded image [%v] in Pod not equal desired cluster image [%v]", pod.Spec.Containers[0].Image, image))
+		}
+	}
+}
+
+func getDockerImageSha() string {
+	sha := tutils.ImageSha
+	if sha != "" {
+		return sha
+	}
+	// If the INFINISPAN_IMAGE_SHA env variable has not been set, attempt to retreive
+	cmd := exec.Command("skopeo inspect docker://quay.io/infinispan/server:12.0 | jq -r '.Digest'")
```

Comment:
@ryanemerson I'm wonder why you are trying to inspect local docker storage for containers which is using inside `kind` since
- `kind` not integrated with local docker registry by default, see https://kind.sigs.k8s.io/docs/user/local-registry/ how to implement this
- `kind` uses `containerd` inside for running images, not `docker` runtime
```
00:37 $ docker ps
CONTAINER ID   IMAGE                   COMMAND                  CREATED       STATUS      PORTS                                                 NAMES
7c4dff1916ca   kindest/node:v1.16.15   "/usr/local/bin/entr…"   10 days ago   Up 4 days   127.0.0.1:37351->6443/tcp, 0.0.0.0:11222->30222/tcp   kind-control-plane
00:37 $ docker exec -it 7c4dff1916ca /bin/bash
root@kind-control-plane:/# ctr --namespace=k8s.io images list | grep infinispan
quay.io/infinispan/server:12.0                                                                    application/vnd.docker.distribution.manifest.v2+json sha256:9c15fbaf399e4087d8cf4e0f98b32f0a30ad9171e0a2f97fe3fba64a3ee93473 192.2 MiB linux/amd64 io.cri-containerd.image=managed 
quay.io/infinispan/server:12.1                                                                    application/vnd.docker.distribution.manifest.v2+json sha256:d9c765412f4b4889457b3e75934137c659057e75d4226cbc04ab76fd9a32a3ad 192.5 MiB linux/amd64 io.cri-containerd.image=managed 
quay.io/infinispan/server@sha256:9c15fbaf399e4087d8cf4e0f98b32f0a30ad9171e0a2f97fe3fba64a3ee93473 application/vnd.docker.distribution.manifest.v2+json sha256:9c15fbaf399e4087d8cf4e0f98b32f0a30ad9171e0a2f97fe3fba64a3ee93473 192.2 MiB linux/amd64 io.cri-containerd.image=managed 
quay.io/infinispan/server@sha256:d9c765412f4b4889457b3e75934137c659057e75d4226cbc04ab76fd9a32a3ad application/vnd.docker.distribution.manifest.v2+json sha256:d9c765412f4b4889457b3e75934137c659057e75d4226cbc04ab76fd9a32a3ad 192.5 MiB linux/amd64 io.cri-containerd.image=managed
```

Reply:
I'm trying to retrieve the sha256 of the remote image, that's why I am installing `skopeo` to perform the inspection.

Reply:
Oh, sure. But you can pull it and use a local inspection to reduce complexity. It the image already using by operator, it can be inspected also.

Reply:
> Oh, sure. But you can pull it and use a local inspection to reduce complexity.

Won't that result in the image being pulled per CI run in the k8 cluster AND the local machine? Either way it still requires the execution of an external process, so I don't think it increases complexibility particularly.

> It the image already using by operator, it can be inspected also.

The image is pulled into the k8 cluster though, not the local repository where the tests are being executed. Do you know how we could get the digest of the image from k8s? Keep in mind that currently no Infinispan cluster is created before the sha is required, so we can't inspect the pods.

Reply:
I think a single way to integrate local docker repository with Kind and pull images before checking sha256.
It can be also load images from the local repository into the `Kind` using 'kind load ...` command
Don't see any other ways and seems to be there is no way to use this kind of tests for both CI and personal (remote OCP/OKD/minikube cluster usage) usage. But need to investigate more.

Reply:
One hacky solution that doesn't involve any external processes and should work on all envs is to do the following before the test begins:

1. Create a standalone pod with the server image
2. Retrieve the image from the spec as it's in the sha256 format
3. Cleanup pod and execute the test